sstables/index_entry.hh

```diff
@@ -191,6 +191,13 @@ public:
     // The caller must ensure that pos remains valid until the future resolves.
     virtual future<std::optional<offset_in_partition>> probe_upper_bound(position_in_partition_view pos) = 0;
 
+    // Tries to warm up cursor's cache so that subsequent advance_to(pos) will hit in the cache.
+    // May do nothing if warming up is not supported by the implementation.
+    // May also improve effectiveness of subsequent probe_upper_bound().
+    // Doesn't affect cursor's position.
+    // pos must be a valid argument to advance_to() in current cursor's state.
+    virtual future<> warm_up(position_in_partition_view pos) = 0;
```

Comment:
I don't understand why we need a warmup; can't the lower bound lookup serve as a warmup?

Also, since we look up in a shared cache now, why would the upper bound lookup fail? We could change it to never initiate I/O, but if I/O is in progress, wait for it. Or launch it after the lower bound lookup completes.

Also #2 (may be a much bigger change and not worth the effort): can we co-lookup the lower and upper bound? Designate the upper bound as opportunistic; all the steps are carried out as usual for the lower bound, and if they happen to also work for the upper bound, great. If they don't, cancel further work for the upper bound.

I guess #2 fails when the lower and upper bounds are reasonable well apart, but the index is well cached.

Reply:
> > I don't understand why we need a warmup; can't the lower bound lookup serve as a warmup?
> 
> We lookup lower bound later in the reader lifecycle, inside fill_buffer(), after the context is created and after we parsed the static row. That's too late. We could advance the cursor here, but that would break the current reader logic, which will attempt to advance the parser later and will think we're already in the right block and fall back to scanning.
> 
> I spent some time investigating this approach, and it can be achieved by changing the logic so that it skips anyway if the current block is ahead. Requires storing current skip_info in the cursor.

I believe the difficulty is due to the evolution of the code, where long ago we did not cache and instead relied on cooperative advancing of the cursors.

Now that we have a cache, can we let the lookups compete? Just make sure to have the loser wait for the winner, so they don't read/parse concurrently.

This will let them run independently, share when they can share, and simplify the logic.

Of course, we might not want to do it in this series, or even at all, 

> 
> > Also, since we look up in a shared cache now, why would the upper bound lookup fail? We could change it to never initiate I/O, but if I/O is in progress, wait for it. Or launch it after the lower bound lookup completes.
> 
> Upper bound lookup never initiates I/O, before and after the patch. It is done after warmup lookup, so there should be no I/O in progress.

I see. This contradicts my earlier competitive lookup comment, since there might not be a competitor.

Still, maybe it's worthwhile to always to the second lookup. Before #1865, the second lookup could have been thrown away, but now it's likely to be valuable when reading the last page. Of course, the user might not want the last page, or we might have a LIMIT, but it seems more valuable now.

> 
> > Also #⁠2 (may be a much bigger change and not worth the effort): can we co-lookup the lower and upper bound? Designate the upper bound as opportunistic; all the steps are carried out as usual for the lower bound, and if they happen to also work for the upper bound, great. If they don't, cancel further work for the upper bound.
> 
> How is it better than what we have here?

Now I'm confused by the code, if we change it, you'll be the one confused.